### PR TITLE
Use inert attribute instead of useDisabled

### PR DESCRIPTION
**Pull request by youknowriad**
## What and Why?

The idea of this PR is that there's a new HTML attribute that was specifically added to address the need of our useDisabled/Disabled hooks/components. So let's just refactor this to use the attribute under the hood and rely on a polyfill for browsers that don't support it yet. The main reason for the switch is that our current useDisabled is too heavy and is buggy in some situations (switching back and forth from disabled to enabled can cause some issues sometimes) 

## Testing Instructions

1- Check that disabled content (block preview, some widget blocks like archives as a dropdown...) are disabled properly 

## Todo

 - [x] Check whether we can deprecate the "Disabled" component.
 - [x] Update existing usage of Disabled with the inert attribute.

### CODE REVIEW

**Comment by youknowriad** [comment]
I'm planning to use the inert attribute to be able to implement the "readonly" view of #44770 without causing a remount of overything.

**Comment by Mamaduka** [comment]
TIL about `inert`.

Do you know if assistive tech can access text in inert elements?

**Comment by youknowriad** [comment]
@Mamaduka No they can't on purpose.

File: packages/compose/README.md [hunk 1]
```diff
@@ -288,11 +288,13 @@ _Returns_
 In some circumstances, such as block previews, all focusable DOM elements
 (input fields, links, buttons, etc.) need to be disabled. This hook adds the
 behavior to disable nested DOM elements to the returned ref.
+If you can, prefer the use of the inert HTML attribute.
```
Comment (ZebulanStanphill) [inline packages/compose/README.md:292]: We should use formatting to make it a bit clear that `inert` is the name of an HTML attribute, rather than merely an adjective. Also, I think there should be another newline here to create a paragraph break.

```suggestion

If you can, prefer the use of the `inert` HTML attribute.
```

File: packages/compose/src/hooks/use-disabled/index.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	return useRefEffect(
+		( node ) => {
+			if ( isDisabledProp ) {
+				return;
+			}
+
+			/** A variable keeping track of the previous updates in order to restore them. */
+			const updates: Function[] = [];
+			const disable = () => {
+				node.childNodes.forEach( ( child ) => {
```
Comment (jorgefilipecosta) [inline packages/compose/src/hooks/use-disabled/index.ts:45]: This only seems to check the direct descents of a node. Should we recursively for each child node also and the inert attribute?

**Review by jorgefilipecosta — APPROVED** [review]
Works well on my tests 👍

File: packages/compose/src/hooks/use-disabled/index.ts (hunk 2 above)
Comment (youknowriad) [reply packages/compose/src/hooks/use-disabled/index.ts:45]: No, inert is recursive by default, the only reason we can't just apply `inert` at the top level is because we don't want to disabled the top level entirely (that's how the hook used to work and we can't change it).

File: packages/block-editor/src/components/block-preview/test/index.js [hunk 3]
```diff
@@ -99,12 +99,6 @@ describe( 'useBlockPreview', () => {
 		);
 		expect( previewedBlockContents ).toBeInTheDocument();
 
-		// Test elements within block contents are disabled.
-		await waitFor( () => {
-			const button = screen.getByText( 'Button' );
-			expect( button.hasAttribute( 'disabled' ) ).toBe( true );
-		} );
```
Comment (ciampo) [inline packages/block-editor/src/components/block-preview/test/index.js:106]: Testing that the preview is disabled seemed like a meaningful test — instead of deleting it, we could have replaced it with user actions? (ie. user tries to interact with block preview, nothing happens)

File: packages/components/src/disabled/test/index.tsx [hunk 4]
```diff
@@ -47,18 +18,14 @@ describe( 'Disabled', () => {
...
 				<Form />
 			</Disabled>
 		);
 
-		const input = screen.getByRole( 'textbox' );
-		const contentEditable = screen.getByTitle( 'edit my content' );
-		expect( input ).toBeDisabled();
-		expect( contentEditable ).toHaveAttribute( 'contenteditable', 'false' );
-		expect( contentEditable ).not.toHaveAttribute( 'tabindex' );
-		expect( contentEditable ).not.toHaveAttribute( 'disabled' );
```
Comment (ciampo) [inline packages/components/src/disabled/test/index.tsx:61]: Same here — instead of testing implementation details, we could write tests from the point of view of the user: _"trying to interact with an element, nothing happens"._

File: packages/components/src/disabled/test/index.tsx [hunk 5]
```diff
@@ -47,18 +18,14 @@ describe( 'Disabled', () => {
...
+		// @ts-ignore
+		expect( container.firstChild.hasAttribute( 'inert' ) ).toBe( true );
```
Comment (ciampo) [inline packages/components/src/disabled/test/index.tsx:28]: Instead of adding `@ts-ignore`, this could have been fixed by adding an optional chaining to `container.firstChild?.hasAttribute( 'inert' )`.

(this line has since been refactored to use the `toHaveAttribute` assertion)

**Review by ciampo — COMMENTED** [review]
This PR seems to have introduced a regression

https://github.com/WordPress/gutenberg/issues/45253

Also, it would be great if we could add CHANGELOG entries and test on Storybook too, while working in the `@wordpress/components` package. Thank you!

File: packages/components/src/disabled/test/index.tsx (hunk 4 above)
Comment (youknowriad) [reply packages/components/src/disabled/test/index.tsx:61]: For me these were testing internal implementation anyway and not good tests so that's why I removed them. That said, I agree that "trying to interact" is a better test here.

Comment (ciampo) [reply packages/components/src/disabled/test/index.tsx:61]: Agreed! The previous tests were definitely testing implementation details too

File: packages/block-library/src/comments/edit/placeholder.js [hunk 6]
```diff
@@ -22,13 +21,8 @@ export default function PostCommentsPlaceholder( { postType, postId } ) {
...
 	);
 
-	const disabledRef = useDisabled();
-
 	return (
-		<div
-			className="wp-block-comments__legacy-placeholder"
-			ref={ disabledRef }
-		>
+		<div className="wp-block-comments__legacy-placeholder" inert="true">
```
Comment (alexstine) [inline packages/block-library/src/comments/edit/placeholder.js:25]: Why exactly are we disabling this?

File: packages/block-library/src/post-comments-form/form.js [hunk 7]
```diff
@@ -13,18 +13,17 @@ import {
...
 
 const CommentsFormPlaceholder = () => {
-	const disabledFormRef = useDisabled();
 	const instanceId = useInstanceId( CommentsFormPlaceholder );
 
 	return (
 		<div className="comment-respond">
 			<h3 className="comment-reply-title">{ __( 'Leave a Reply' ) }</h3>
-			<form noValidate className="comment-form" ref={ disabledFormRef }>
+			<form noValidate className="comment-form" inert="true">
```
Comment (alexstine) [inline packages/block-library/src/post-comments-form/form.js:26]: Why is this added to an entire form?

File: packages/block-library/src/table-of-contents/edit.js [hunk 8]
```diff
@@ -295,7 +293,7 @@ export default function TableOfContentsEdit( {
 	return (
 		<>
 			<nav { ...blockProps }>
-				<ol ref={ disabledRef }>
+				<ol inert="true">
```
Comment (alexstine) [inline packages/block-library/src/table-of-contents/edit.js:296]: As noted here, this introduced an accessibility regression. Please don't do this in the future.

https://github.com/WordPress/gutenberg/pull/54322#discussion_r1320386448

**Review by alexstine — COMMENTED** [review]
@youknowriad Please see my comments below. I fear using `inert` has caused quite a few accessibility regressions. I am not sure why for example you added it to the table of contents block, this made the block content that sighted users can see totally inaccessible to keyboard users on the back-end. As noted by @Mamaduka, further research probably should have happened here before shipping this PR.

File: packages/block-library/src/comments/edit/placeholder.js (hunk 6 above)
Comment (youknowriad) [reply packages/block-library/src/comments/edit/placeholder.js:25]: This is just mirrors what existed before the PR. Basically we had a useDisabled hook that poly fills what "inert" does. So I'm not sure we changed anything in behavior here. I believe these "inert" within components are there because these blocks are "interactive", people can interact with them in the frontend but they are not meant to be interacted with in the backend (or navigated within). I'm not entirely sure that we really want to keep this visible for screen reader users, they are just mean to offer a visual representation of what the block will look like in the frontend but for interactions purposes, the intent is to hide these entirely. 

If we really want to leave a "trace", we may add a label maybe or something like that?

Comment (alexstine) [reply packages/block-library/src/comments/edit/placeholder.js:25]: @youknowriad That's the problem. We are giving sighted people information keyboard/screen reader users don't have access to. If the content on the back-end is really not important, let's add `display: none;` to it so it is inaccessible to everyone. Additional context make sense?

Thanks.

Comment (youknowriad) [reply packages/block-library/src/comments/edit/placeholder.js:25]: I think I see what you're coming from but the reality is that the editor is also meant to be WYSIWYG and that concept doesn't really translate into screen readers. Sighted users will see something different visually between the backend and frontend and consider this a visual bug. 

In terms of interactions, I agree, both sighted and screen reader users should have the exact same thing but the visuals are important for sighted users.

Would it help if we add a label or something saying "Visual Preview of block/whatever we're previewing" or something like that?

---

Also, this is independent of the use of inert, it has always been this way in the editor.

Comment (alexstine) [reply packages/block-library/src/comments/edit/placeholder.js:25]: @youknowriad 

> Also, this is independent of the use of inert, it has always been this way in the editor.

That's got to change but I guess it can go into an issue for further discussion.

Labels don't help here because the preview itself is still not communicated equally.

Why is it not possible to disable interactions with blocks in an accessible way? #54322 accomplishes just that. People with vision can't interact with the block and screen readers can still see the block content but cannot interact with the block.

We've got to quit caring so much about this front-end vs. back-end. Honestly, do you think anyone at all would notice the usage of `inert` vs. `aria-disabled` with event listeners to disable a control? Absolutely not. It just makes it more work for developers and that I fear is the real problem here.

Comment (alexstine) [reply packages/block-library/src/comments/edit/placeholder.js:25]: Below example shows how this can still remain accessible but preventing the form from actually working.

```html
<html>
	<head>
		<title>Test</title>
	</head>
	<body>
		<h1>With inert</h1>
		<form class="form1" action="https://www.google.com" method="GET" inert>
			<input type="search" value="" name="search1" aria-label="Search 1" placeholder="Search 1" />
			<input type="submit" value="Search 1" />
		</form>
		<br />
		<h1>Accessible way but more work</h1>
		<form class="form2" action="https://www.google.com" method="GET">
			<input type="search" value="" name="search2" aria-label="Search 2" placeholder="Search 2" aria-disabled="true" />
			<input type="submit" value="Search 2" aria-disabled="true" />
		</form>
		<script type="text/javascript">
			const form2 = document.querySelector('.form2');
			form2.addEventListener('submit', function(e) {
				e.preventDefault();
			} );
		</script>
	</body>
</html>
```

Comment (ramonjd) [reply packages/block-library/src/comments/edit/placeholder.js:25]: I've created an issue where can record discussions and debate alternatives:

- https://github.com/WordPress/gutenberg/issues/54369

Comment (youknowriad) [reply packages/block-library/src/comments/edit/placeholder.js:25]: Now that I think about it, all of our usage of "inert" can be seen as "screenshots". In that sense, I think maybe the solution going forward is to use a role="img" on these elements with a label. When you show a preview of a block or a template, it's just like an image.
